mvlearn/embed/kcca.py

```diff
@@ -105,328 +161,118 @@ def fit(self, Xs):
-
-    # Allocate left-hand side (LH) and right-hand side (RH):
-    LH = np.zeros((sum(nFs), sum(nFs)))
-    RH = np.zeros((sum(nFs), sum(nFs)))
-
-    # Fill the left and right sides of the eigenvalue problem
-    for i in range(nDs):
-        RH[
-            sum(nFs[:i]): sum(nFs[: i + 1]), sum(nFs[:i]): sum(nFs[: i + 1])
-        ] = crosscovs[i * (nDs + 1)] + reg * np.eye(nFs[i])
-
-        for j in range(nDs):
-            if i != j:
-                LH[
-                    sum(nFs[:j]): sum(nFs[: j + 1]),
-                    sum(nFs[:i]): sum(nFs[: i + 1])
-                ] = crosscovs[nDs * j + i]
-
-    LH = (LH + LH.T) / 2.0
-    RH = (RH + RH.T) / 2.0
-
-    maxCC = LH.shape[0]
-    r, Vs = eigh(LH, RH, eigvals=(maxCC - n_components, maxCC - 1))
-    r[np.isnan(r)] = 0
-    rindex = np.argsort(r)[::-1]
-    comp = []
-    Vs = Vs[:, rindex]
-    for i in range(nDs):
-        comp.append(Vs[sum(nFs[:i]): sum(nFs[: i + 1]), :n_components])
-    return comp
-
-
-def _zscore(d):
-    """
-    Calculates z-score of data
+        return self.components_
 
-    Parameters
-    ----------
-    d : array
-        Data of interest
 
-    Returns
-    -------
-    z : array
-        Z-score
-    """
-    z = (d - d.mean(0)) / d.std(0)
-    return z
+def _center_norm(x):
+    N = len(x)
+    x = x - numpy.matlib.repmat(np.mean(x, axis=0), N, 1)
+    return x@np.sqrt(np.diag(np.divide(1, np.diag(np.transpose(x)@x))))
 
 
-def _listdot(d1, d2):
-    """
-    Calculates the dot product between two arrays
+def _make_kernel(x, ktype, constant=10, degree=2.0, sigma=1.0):
+    N = len(x)
+    N0 = np.eye(N)-1/N*np.ones((N, N))
 
-    Parameters
-    ----------
-    d1 : array
-         Data of interest
-    d1 : array
-         Data of interest
-
-    Returns
-    -------
-    ld : list
-        Dot product
-    """
-    ld = [np.dot(x[0].T, x[1]) for x in zip(d1, d2)]
-    return ld
-
-
-def _listcorr(a):
-    """
-    Returns pairwise row correlations for all items
-    in array as a list of matrices
-
-    Parameters
-    ----------
-    a : list of array-likes
-
-    Returns
-    -------
-    corrs_ : list of array-likes
-             Pairwise row correlations for all items in array
-    """
-    corrs_ = np.zeros((a[0].shape[1], len(a), len(a)))
-    for i in range(len(a)):
-        for j in range(len(a)):
-            if j > i:
-                corrs_[:, i, j] = [
-                    np.nan_to_num(np.corrcoef(ai, aj)[0, 1])
-                    for (ai, aj) in zip(a[i].T, a[j].T)
-                ]
-    return corrs_
-
-
-def _rowcorr(a, b):
-    """
-    Finds correlations between corresponding matrix rows (a and b)
-
-    Parameters
-    ----------
-    a : array
-        Matrix row 1
-    b : array
-        Matrix row 2
-
-    Returns
-    -------
-    cs: array
-        Correlations between corresponding rows
+    # Linear kernel
+    if ktype == "linear":
+        return N0@(x@x.T + constant)@N0
```

Comment:
I think that the constant is problematic and maybe shouldn't be a parameter (always 0). By setting it large relative to teh input data, like the tutorials, it becomes the only relevant term and so you get perfect correlation which isn't useful.

Reply:
i see what youre saying. will remove the constant parameter